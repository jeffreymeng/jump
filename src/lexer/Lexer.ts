import Token, { TOKEN_TYPE } from "./Token";
import SourcePosition from "./SourcePosition";
import Source from "./Source";

/**
 * An error that occurred while lexing, typically not due to the input but due to
 * an implementation bug.
 */
export class LexerError extends Error {
	public readonly position: SourcePosition;

	constructor(message: string, position: SourcePosition) {
		super(message);
		this.name = "Lexer Error";
		this.position = position;
	}
}

/**
 * "SheetsScript" syntax error -- so it's not the same as a native syntax error.
 */
export class SSSyntaxError extends Error {
	public readonly position: SourcePosition;

	constructor(message: string, position: SourcePosition) {
		super(message);
		this.name = "SSSyntax Error";
		this.position = position;
	}
}


export default class Lexer implements Iterator<Token> {
	/**
	 * The source code of the lexer.
	 */
	public readonly source: Source;

	/**
	 * The index of the current character within the source.
	 */
	protected index = 0;

	/**
	 * The index of the current token.
	 */
	protected currentTokenIndex = 0;

	constructor(source: Source) {
		this.source = source;
	}

	public static readonly DIGITS = "0123456789";



	/**
	 * Consumes the next token, and returns it.
	 */
	public nextToken(): Token {
		const c = this.source.next();
		this.setTokenIndex();

		if ("\"'".includes(c)) {
			// consume until we see the same type of quote again
			const buffer = this.source.consumeUntil(c);
			return new Token(TOKEN_TYPE.STRING, buffer);
		} else if (Lexer.DIGITS.includes(c)) {
			let buffer = c + this.source.consumeAll(Lexer.DIGITS);
			// if the next character is a dot, then this is part of a float, so keep going.
			if (this.source.peekNext() === ".") {
				buffer += this.source.next() + this.source.consumeAll(Lexer.DIGITS);
			}
			return new Token(TOKEN_TYPE.NUMBER, buffer);
		} else if (c === "\n") {
			return new Token(TOKEN_TYPE.CONTROL, "\n");
		} else if (/[ \t]/g.test(c)) {
			// skip any spaces and tabs.
			return this.nextToken();
		}

		throw new SSSyntaxError(
			`Unexpected character '${c}'.`,
			this.source.getPosition()
		);
	}

	public hasNextToken(): boolean {
		// if there are only whitespace characters (including newlines)
		// then we can
		return !/[\s\n]*/g.test(this.source.peekNext());
	}

	public next(): IteratorResult<Token> {
		if (!this.hasNextToken()) {
			return {
				value: null,
				done: true
			}
		}
		return {
			value: this.nextToken(),
			done: this.hasNextToken(),
		};
	}

	/**
	 * Set the current token position to the lexer's next character's position
	 */
	protected setTokenIndex() {
		this.currentTokenIndex = this.index;
	}
}
