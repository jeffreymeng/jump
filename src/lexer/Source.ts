import SourcePosition from "./SourcePosition";
import { JumpSyntaxError } from "./Lexer";

export default class Source {
	protected _index = 0;
	public readonly text: string;

	/**
	 * Returns the current index of the source.
	 */
	public get index() {
		return this._index;
	}

	constructor(text: string) {
		this.text = text;
	}

	/**
	 * Returns the next character from the source, updating the line and position as necessary.
	 * If no next character exists, returns an empty string.
	 */
	public next(): string {
		if (this._index >= this.text.length) {
			throw new JumpSyntaxError(
				"Unexpected end of input.",
				new SourcePosition(this.text, this._index)
			);
		}
		const c = this.text[this._index];
		this._index++;
		return c;
	}

	/**
	 * Returns the next character from the source without changing the current position.
	 * @param offset - If provided, the ith character after the current one is returned.
	 * If `i` is out of range, an empty string will be returned.
	 */
	public peekNext(offset = 0): string {
		if (this._index + offset >= this.text.length) {
			return "";
		}

		return this.text[this._index + offset];
	}

	/**
	 * Returns the next `len` characters from the source without changing the current position.
	 * @param len - The number of characters to return, or `0` to return the rest of source. Must be in range
	 */
	public peekNextCharacters(len = 0): string {
		if (len === 0) {
			return this.text.substring(this._index);
		}

		if (this._index + len >= this.text.length) {
			throw new JumpSyntaxError(
				"Unexpected end of file",
				new SourcePosition(this.text, this._index)
			);
		}
		return this.text.substring(this._index, this._index + len);
	}

	/**
	 * Consumes characters until any character inside endChars is found.
	 * The end character is consumed, but only the characters from
	 * the current position (inclusive) to the last character before
	 * the end character are returned. The string must eventually contain
	 * one of the endChars.
	 * @param endChars - A string containing one or more characters that will cause consumeUntil()
	 *  to stop consuming characters.
	 * @param errorMessageIfNotFound - If the string does not eventually contain
	 * an endChar, then this message will be used instead of the default error message.
	 */
	public consumeUntil(
		endChars: string,
		errorMessageIfNotFound = "Unexpected end of input."
	): string {
		const consumed = this.consumeWhile((c) => !endChars.includes(c));
		// consume the character that caused the predicate to fail, unless
		// we reached the end of the string already (and thus nothing caused
		// this to throw), which would cause an error.
		if (this._index < this.text.length) {
			this.next();
		} else {
			throw new JumpSyntaxError(
				errorMessageIfNotFound,
				this.getPosition()
			);
		}
		return consumed;
	}

	/**
	 * Consumes and returns characters until a character not in `chars` is found.
	 * @param chars - A string containing one or more characters to be consumed continuously.
	 */
	public consumeAll(chars: string): string {
		return this.consumeWhile((c) => chars.includes(c));
	}

	/**
	 * Consumes characters until the predicate is no longer true.
	 * All characters from the starting position until the predicate fails will be consumed and returned.
	 * The character that failed the predicate will NOT be consumed.
	 * @param predicate - A function returning true if a character passes the predicate.
	 */
	public consumeWhile(predicate: (c: string) => boolean): string {
		let buffer = "";
		while (this.hasNext() && predicate(this.peekNext())) {
			buffer += this.next();
		}
		return buffer;
	}

	/**
	 * Get a SourcePosition object corresponding to the `i`th character of the source,
	 * or the current index if none is provided.
	 */
	public getPosition(i = this._index) {
		return new SourcePosition(this.text, i);
	}

	/**
	 * Returns true if the source has a next token (e.g. has not reached the end yet).
	 */
	public hasNext() {
		return this._index < this.text.length;
	}
}
