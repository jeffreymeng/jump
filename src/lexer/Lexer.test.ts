import { describe, test } from "@jest/globals";

import Lexer from "./Lexer";
import Source from "./Source";

describe("Lexer", () => {
	test("Basic Lexing", () => {
		const s = `"hello" 2 3 0.2 42.294 "hello" "23.1"`;
		const lexer = new Lexer(new Source(s));
		// eventually there should be an error
		for (const token of lexer) {
			console.log(token);
		}
	});
});
