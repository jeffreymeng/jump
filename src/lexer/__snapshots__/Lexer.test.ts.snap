// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Lexer Error cases 1`] = `"Unexpected character 'â€¢'."`;

exports[`Lexer Literals 1`] = `"Expected string to be closed."`;

exports[`Lexer Literals 2`] = `"Expected string to be closed."`;

exports[`Lexer Literals 3`] = `"Expected string to be closed."`;
